src/vision/status_quo/barbara_battles_buffered_streams.md

```diff
@@ -0,0 +1,216 @@
+  --> src/main.rs:8:18
+   |
+8  | async fn do_work(database: &Database) {
+   |                  ^^^^^^^^  --------- this data with an anonymous lifetime `'_`...
+   |                  |
+   |                  ...is captured here...
+   |        .for_each(|work_item| task::spawn(async move {
+   |                              ----------- ...and is required to live as long as `'static` here
+```
+
+"Ah, right," she says, "spawned tasks can't use borrowed data. I wish I had [rayon] or the scoped threads from [crossbeam]." (What Barbara doesn't realize is that spawning wouldn't actually have fixed her problem anyway: the `for_each` combinator would have awaited the resulting `JoinHandle` and hence it would have blocked... but she could have tweaked her program to fix that if she had gotten that far.)
+
+"Let me see," Barbara thinks. "What else could I do?" She has the idea that she doesn't have to process the work items immediately. She could buffer up the work into a [`FuturesUnordered`] and process it after everything is ready:
+
+```rust
+async fn do_work(database: &Database) {
+    let work = do_select(database, FIND_WORK_QUERY)?;
+    let mut results = FuturesUnordered::new();
+    stream::iter(work.into_iter())
+        .map(|item| do_select(database, work_from_item(item)))
+        .buffered(5)
+        .for_each(|work_item| {
+            results.push(process_work_item(database, work_item));
+            futures::future::ready(())
+        })
+        .await;
+
+    while let Some(_) = results.next().await { }
+}
+```
+
+This isn't maximally efficient -- it's introducing an arbitrary phasing into her work -- but at least it doesn't cause timeouts. Going forward, she tries to remember never to do a m"nested `await`" like this. Buffering up work into a `FuturesUnordered` becomes a pattern she finds herself using throughout the codebase.
+
+## 🤔 Frequently Asked Questions
+
+### **What are the morals of the story?**
+
+* Rust's future model is a 'leaky abstraction' that works quite differently from futures in other languages. It is prone to some subtle bugs that require a relatively deep understanding of its inner works to understand and fix.
+* "Nested awaits" -- where the task blocks on an inner await while there remains other futures that are still awaiting results -- are easy to do but can cause a lot of trouble.
+* Lack of scoped futures makes it hard to spawn items into separate tasks for independent processing sometimes.
+
+### **What are the sources for this story?**
+
+This is based on the bug report [Footgun with Future Unordered](https://github.com/rust-lang/futures-rs/issues/2387) but the solution that Barbara came up with is something that was relayed by [farnz](https://github.com/farnz) vision doc writing session. [farnz] mentioned at the time that this pattern was frequently used in their codebase to work around this sort of hazard.
```

Comment:
We don't use this pattern frequently - it's a hack to handle places where refactoring so that the `FuturesUnordered` just polls `JoinHandle`s is hard.

The normal pattern is to refactor `do_select`:

```
pub async fn do_select(...) -> Result<_, anyhow::Error> {
    // Do anything that depends on borrows from the arguments - noting that we can't `.await` at this point.
    let query = format_query!(SELECT_QUERY, ...);
    // Then spawn off a block that has 'static lifetime and does the real work, with `.await` as needed.
    tokio::spawn(async move {
        ...
    }).await?
}
```
but we use the other pattern for cases where we need to `.await` before finishing with all our borrows.

Also worth noting that the documented pattern here that I mentioned in the vision session radically changes behaviour - the solution that gets the behaviour we intended (at most 5 work handles dequeued by `do_select`, at most one work item in flight with `process_work_item`) is the spawn inside `do_select`

Reply:
Ah, thanks! This is super helpful! I wasn't really sure what behavior you were shooting for.